### PR TITLE
Skip upload of segments_N file

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- Currently, while uploading segments to remote segment store, we upload existing `segments_N` file as well.
- But we don't use the `segments_N` file in the download flow as we commit `SegmentInfosSnapshot` that is stored in the form of bytes in remote segment metadata.
- In this change, we skip uploading the `segments_N` file.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/9433

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -4740,36 +4738,30 @@ public void syncSegmentsFromGivenRemoteSegmentStore(
...
+            // delete any other commits, we want to start the engine only from a new commit made with the downloaded infos bytes.
+            // Extra segments will be wiped on engine open.
+            for (String file : List.of(store.directory().listAll())) {
+                if (file.startsWith(IndexFileNames.SEGMENTS)) {
+                    store.deleteQuiet(file);
                 }
             }
+            assert Arrays.stream(store.directory().listAll()).filter(f -> f.startsWith(IndexFileNames.SEGMENTS)).findAny().isEmpty()
+                : "There should not be any segments file in the dir";
+            store.commitSegmentInfos(infosSnapshot, processedLocalCheckpoint, processedLocalCheckpoint);
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:4764]: This still needs some re-factoring as a lot of code is duplicated. Will get it fixed in next commit

**Review by Bukhtawar — COMMENTED** [review]
Curious would segments_N be needed by other flows like CCR?

**Comment by mch2** [comment]
I think this is a good simplification but would like to clarify the behavior for restore within `syncSegmentsFromGivenRemoteSegmentStore` where a pterm/gen is specified.  This method states `* Downloads segments from given remote segment store for a specific commit`.  Would this now fetch the latest checkpoint within the commit gen?  Or did it already do this by invoking `listFilesByPrefixInLexicographicOrder` ?

> Curious would segments_N be needed by other flows like CCR?

I would think CCR restore flow we'd want to reuse segrep mechanisms used within a cluster, which do not depend on the commit file.

**Comment by mch2** [comment]
This will also resolve the flakiness with https://github.com/opensearch-project/OpenSearch/issues/8059 - where it is failing because the listener is trying to upload a segments_n deleted on disk by a concurrent flush.

**Comment by sachinpkale** [comment]
Looking into failed tests

**Comment by ticheng-aws** [comment]
Hi @sachinpkale, the PR is stalled. Is this being worked upon? Feel free to reach out to maintainers for further reviews.

**Comment by sachinpkale** [comment]
Not going ahead with this change as of now. We will revisit this once we get rid of explicit flush for segment replication + remote store.

**Pull request closed without merging**